app/models/email_token.rb

```diff
@@ -1,97 +1,107 @@
 # frozen_string_literal: true
 
 class EmailToken < ActiveRecord::Base
+  class TokenAccessError < StandardError; end
+
   belongs_to :user
 
-  validates :token, :user_id, :email, presence: true
+  validates :user_id, :email, :token_hash, presence: true
 
-  before_validation(on: :create) do
-    self.token = EmailToken.generate_token
-    self.email = self.email.downcase if self.email
+  scope :unconfirmed, -> { where(confirmed: false) }
+  scope :active, -> { where(expired: false).where('created_at >= ?', SiteSetting.email_token_valid_hours.hours.ago) }
+
+  after_initialize do
+    if self.token_hash.blank?
+      @token ||= SecureRandom.hex
+      self.token = @token
+      self.token_hash = self.class.hash_token(@token)
+    end
   end
 
   after_create do
-    # Expire the previous tokens
-    EmailToken.where(user_id: self.user_id)
-      .where("id != ?", self.id)
+    EmailToken
+      .where(user_id: self.user_id)
+      .where(scope: [nil, self.scope])
+      .where.not(id: self.id)
       .update_all(expired: true)
   end
 
-  def self.token_length
-    16
-  end
-
-  def self.valid_after
-    SiteSetting.email_token_valid_hours.hours.ago
+  before_validation do
+    self.email = self.email.downcase if self.email
   end
 
-  def self.unconfirmed
-    where(confirmed: false)
+  before_save do
+    if self.scope.blank?
+      STDERR.puts "EmailToken has an empty scope"
```

Comment:
Debug leftover?

Reply:
Yes and no. I will leave a deprecation notice here to make sure that `scope` is always filled.

Reply:
- [x] The backbone plugins will continue inline I'm sure you can do ![image](https://user-images.githubusercontent.com/76078201/144322864-3d80c80c-d760-4350-90bd-35c90a496d1e.jpeg)